Reduce use of WTF_ALLOW_UNSAFE_BUFFER_USAGE in CSSSelectorList

cdumez opened the pull request:
#### 028ef08f4dbde1d974f45a43d9c6816d492724d4
<pre>
Reduce use of WTF_ALLOW_UNSAFE_BUFFER_USAGE in CSSSelectorList
<a href="https://bugs.webkit.org/show_bug.cgi?id=304791">https://bugs.webkit.org/show_bug.cgi?id=304791</a>

Reviewed by Sam Weinig.

Reduce use of WTF_ALLOW_UNSAFE_BUFFER_USAGE in CSSSelectorList. This
tested as performance neutral on Speedometer.

* Source/WebCore/css/CSSSelectorList.cpp:
(WebCore::CSSSelectorList::CSSSelectorList):
(WebCore::CSSSelectorList::makeCopyingSimpleSelector):
(WebCore::CSSSelectorList::makeCopyingComplexSelector):
(WebCore::CSSSelectorList::makeJoining):
(WebCore::CSSSelectorList::componentCount const):
(WebCore::CSSSelectorList::listSize const):
* Source/WebCore/css/CSSSelectorList.h:
(WebCore::CSSSelectorList::CSSSelectorList):
(WebCore::CSSSelectorList::isEmpty const):
(WebCore::CSSSelectorList::indexOfNextSelectorAfter const):
* Source/WebCore/css/StyleRule.cpp:
(WebCore::StyleRule::createForSplitting):

Canonical link: <a href="https://commits.webkit.org/305131@main">https://commits.webkit.org/305131@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c0b8193f3f02d6a7bfc4873a5c81ae7e3f2de925

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/137604 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/9965 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/48892 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/145362 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/90577 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/df1d3fec-e547-4a8f-a533-e5adddd22252) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/139476 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/10668 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/10095 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/105240 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/76834 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/0cee9bec-d3de-4409-9998-81c351aa2cc0) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/140549 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/7923 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/123319 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/86096 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/b021dfce-43cf-4af8-9589-548cb57e1a2f) 
| | [⏳ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/iOS-26-Simulator-WPT-WK2-Tests-EWS "Waiting to run tests") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/165/builds/5277 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/5945 "Built successfully") | | 
| | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/41485 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/148129 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/9647 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/42030 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/113634 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/9665 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/8132 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/113966 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28925 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/7483 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/119559 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/64311 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/9696 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/37609 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/9427 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/73261 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/9636 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/9488 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

weinig reviewed (commented):
If we switch CSSSelectorList to a FixedVector, we can also probably get rid of all the complexity with m_isLastInSelectorList and the complexity in CSSSelectorList::componentCount() (since it will just be the fixed vector's size).

I am curious about the history here, and why such lengths were gone to avoid storing a length for these. Premature size optimization? Something else?

weinig commented on Source/WebCore/css/CSSSelectorList.cpp:
```diff
@@ -59,18 +51,18 @@ CSSSelectorList::CSSSelectorList(MutableCSSSelectorList&& selectorVector)
                 // Move item from the parser selector vector into m_selectorArray without invoking destructor (Ugh.)
                 auto* currentSelector = current->releaseSelector().release();
+WTF_ALLOW_UNSAFE_BUFFER_USAGE_BEGIN
                 memcpy(static_cast<void*>(&m_selectorArray[arrayIndex]), static_cast<void*>(currentSelector), sizeof(CSSSelector));
```
Wow, I wonder if this is pre-move semantics and just be a proper move (and added move-constructors to CSSSelector) nowadays. (Obviously not needed for this change).

weinig commented on Source/WebCore/css/CSSSelectorList.cpp:
```diff
@@ -131,17 +118,15 @@ CSSSelectorList CSSSelectorList::makeJoining(const CSSSelectorList& a, const CSS
     for (size_t i = 0; i < aComponentCount; ++i)
         new (NotNull, &selectorArray[i]) CSSSelector(a.m_selectorArray[i]);
     for (size_t i = 0; i < bComponentCount; ++i)
         new (NotNull, &selectorArray[aComponentCount + i]) CSSSelector(b.m_selectorArray[i]);
```
This initialization could use FixedVector<CSSSelector>::createWithSizeFromGenerator() to get the same behavior.

cdumez replied:
I am not sure how to use `createWithSizeFromGenerator()` here. The original code was using placement new. Did you mean this?
```
    auto selectorArray = FixedVector<CSSSelector>::createWithSizeFromGenerator(aComponentCount + bComponentCount, [&](size_t i) {
        return i < aComponentCount ? a.m_selectorArray[i] : b.m_selectorArray[i - aComponentCount];
    });
```

weinig replied:
Yeah. If you trace that down to the TrailingArray constructor that takes the generator, you will see it does the placement new.

cdumez replied:
Ok. I'll re-do Speedometer A/B testing with the couple of changes you recommended just to be safe.

weinig commented on Source/WebCore/css/StyleRule.cpp:
```diff
@@ -332,12 +332,10 @@ void StyleRuleWithNesting::wrapperAdoptOriginalSelectorList(CSSSelectorList&& se
-WTF_ALLOW_UNSAFE_BUFFER_USAGE_BEGIN
+    FixedVector<CSSSelector> selectorListArray(selectors.size());
     for (unsigned i = 0; i < selectors.size(); ++i)
         new (NotNull, &selectorListArray[i]) CSSSelector(*selectors.at(i));
```
This initialization could use FixedVector<CSSSelector>::map() to get the same behavior.

weinig replied:
I might also add a CSSSelectorList constructor that takes a `const Vector<const CSSSelector*>& selectors` to avoid having this knowledge of the internals in StyleRule.

cdumez replied:
Will add the new constructor you suggested via https://github.com/WebKit/WebKit/pull/56132

weinig commented on Source/WebCore/css/CSSSelectorList.h:
```diff
@@ -43,21 +43,19 @@ class CSSSelectorList {
-    bool isEmpty() const { return !m_selectorArray; }
-    const CSSSelector* first() const LIFETIME_BOUND { return m_selectorArray.get(); }
+    bool isEmpty() const { return m_selectorArray.isEmpty(); }
+    const CSSSelector* first() const LIFETIME_BOUND { return m_selectorArray.begin(); }
```
Is this using m_selectorArray.begin() rather than &m_selectorArray.first() to make it work with an empty list? If so, I think a comment would probably be worthwhile, since that isn't a common idiom in the code base.

cdumez committed:
Reduce use of WTF_ALLOW_UNSAFE_BUFFER_USAGE in CSSSelectorList
https://bugs.webkit.org/show_bug.cgi?id=304791

Reviewed by Sam Weinig.

Reduce use of WTF_ALLOW_UNSAFE_BUFFER_USAGE in CSSSelectorList. This
tested as performance neutral on Speedometer.

* Source/WebCore/css/CSSSelectorList.cpp:
(WebCore::CSSSelectorList::CSSSelectorList):
(WebCore::CSSSelectorList::makeCopyingSimpleSelector):
(WebCore::CSSSelectorList::makeCopyingComplexSelector):
(WebCore::CSSSelectorList::makeJoining):
(WebCore::CSSSelectorList::componentCount const):
(WebCore::CSSSelectorList::listSize const):
* Source/WebCore/css/CSSSelectorList.h:
(WebCore::CSSSelectorList::CSSSelectorList):
(WebCore::CSSSelectorList::isEmpty const):
(WebCore::CSSSelectorList::indexOfNextSelectorAfter const):
* Source/WebCore/css/StyleRule.cpp:
(WebCore::StyleRule::createForSplitting):

Canonical link: https://commits.webkit.org/305131@main
